invoke-ai/InvokeAI
# feat(db): add SQLiteMigrator to perform db migrations

## What type of PR is this? (check all applicable)


## Have you discussed this change with the InvokeAI team?


## Description

This PR enhances our SQLite database with migration logic.

### `SQLiteMigrator` class

The new `SQLiteMigrator` class handles safely running database migrations. It is initialized in the `SqliteDatabase` class's init, and immediately runs all database migrations.

### `Migration` class

Migrations are reprsented by a `Migration` class, which has 3 attributes:

- `db_version: int`: The database version this migration results in.
- `app_version: str`: The semver app version this migration is run for.
- `migrate: Callable[[sqlite3.Cursor], None]`: A function that performs the migration. It receives a cursor _only_, but can do anything it wants to do. A convention is established for these functions.

All schema-creating SQL now lives in a `migrate` function. We haven't needed to make any data migrations yet, but when we do, this will also be handled within one of these callbacks.

### Migration Flow

First, migrations are registered with `SQLiteMigrator` with it's `register_migration` method. This performs some basic checks of the migration version.

After registering all migrations, they are run with the `run_migrations` method. This does a few things:

- Creates a `version` table in the DB, if it doesn't already exist. This table has `db_version INTEGER`, `app_version TEXT` and `migrated_at DATETIME` columns.
- Sort the migrations by their `db_version`.
- Do some checks to see if we need a migration.
- Backs up the database (if it's a file database). The migration bails out if this fails.
- Runs each migration. If there is a problem, restore from backup.

### Included Migrations

Migrations are in `invokeai/app/services/shared/sqlite/migrations`.

#### `migrate_1.py`

All\* schema SQL up to 3.4.0post2 is in `migration_1.py`. Running only this migration should result in a database that is identical to the one you get from starting up 3.4.0post2.

SQL in this migration is **idempotent** (same as it was when the SQL was spread across the various services).

#### `migrate_2.py`

Schema changes through 3.5.0 (the upcoming release) are in `migration_2.py`.

SQL in this migration is **not idempotent**. Future migrations need not be idempotent, as the migration logic ensures each will only be run once.

### \*Caveat - ItemStorage

This class provides a generic document-db-like interface for storing objects. Our `graph_executions` and `graphs` tables are created and managed by this service. This PR does not touch this class and therefore does not touch either of those two tables.

We can decide how to handle those tables in the future as the need arises.

### Change to Model Manager Metadata table

I noticed that there is a `model_manager_metadata` table which included the app version, and whose `version` property wasn't accessed outside the service.

I believe the new `version` table fulfills the purpose of this table, and have removed it.

@lstein Please let me know if this is not right.

## QA Instructions, Screenshots, Recordings

1.  Case 1 - Upgrade

    - Back up your 3.4.0post2 database
    - Run this PR
    - It should upgrade your database and everything should work exactly like it did before

2.  Case 2 - New Install

    - Move your database out of the invoke root so that when the app starts, it creates a new one
    - Run this PR
    - It should work just like a new install

3.  Case 3 - With an In-Memory Database

    - Enable the in-memory memory database (set `use_memory_db` under `Paths` in `invokeai.yaml` to `true`)
    - Run this PR
    - It should work just like a new install


## Added/updated tests?

- [x] Yes: Fairly comprehensive tests are added for the `SQLiteMigrator`.
      have not been included_
